txscript/script.go

```diff
@@ -281,20 +281,52 @@ func removeOpcodeByData(pkscript []parsedOpcode, data []byte) []parsedOpcode {
 		}
 	}
 	return retScript
-
 }
 
-// CalcSignatureHash is an exported version for testing.
-func CalcSignatureHash(script []parsedOpcode, hashType SigHashType,
-	tx *wire.MsgTx, idx int, cachedPrefix *chainhash.Hash) ([]byte, error) {
-	return calcSignatureHash(script, hashType, tx, idx, cachedPrefix)
+// optimizedWitnessSigningHash is a fast calculation of the witness for
+// signing to be used when signing with the SigHashAll optimization flag on.
+// It writes the serialized pkScript as it would be in the form of a serialized
+// msgTx with a witness signing version to a buffer, then hashes that.
+func optimizedWitnessSigningHash(script []parsedOpcode, hashType SigHashType,
+	tx *wire.MsgTx, idx int) (chainhash.Hash, error) {
+	var buf bytes.Buffer
+	if hashType&sigHashMask != SigHashAllValue {
+		binary.Write(&buf, binary.LittleEndian,
```

Comment:
Please avoid binary.Write, especially since this is supposed to be an optimized hot path.  Call the methods on binary.LittleEndian instead.